contracts/specs/NFTSwapper/NFTSwapperSpec.md

```diff
@@ -0,0 +1,163 @@
+# NFTSwapper Spec
+
+NFTSwapper enables permissionless swapping between ERC721's within a collection. 
+
+This contract is immutable with no owner. No NFTs or ETH are ever held by this contract, instead it just facilitates the permissionless swap process. 
+
+---
+
+- [NFTSwapper Spec](#nftswapper-spec)
+  - [Function Interface](#function-interface)
+    - [`constructor`](#constructor)
+    - [`swapPrice`](#swapprice)
+    - [`swap`](#swap)
+  - [Swap Price Curve](#swap-price-curve)
+    - [Constant Price Curve](#constant-price-curve)
+    - [Output NFT age based curve](#output-nft-age-based-curve)
+    - [Input and Output NFT age difference based curve](#input-and-output-nft-age-difference-based-curve)
+  - [Approval Considerations](#approval-considerations)
+
+## Function Interface
+
+### `constructor`
+
+```solidity
+constructor(
+    address swapPool, 
+    address nft, 
+    address feeRecipient, 
+    PriceCurveParams priceCurveParams
+)
+```
+
+* `swapPool`: the address which holds the swappable nft's, note that the `swapPool` must approve this contract for it's NFT's (or a subset of them, see [Approval Considerations](#approval-considerations)).
+* `nft`: the address of the ERC721 token that is being swapped.
+* `feeRecipient`: the address where swap fees will be sent. 
+* `priceCurveParams`: the parameters for the price curve (see [Swap Price Curve](#swap-price-curve)).
+ 
+### `swapPrice`
+
+```solidity
+function swapPrice(
+    uint256 inputTokenId, 
+    uint256 outputTokenId
+) pure returns (uint256)
+```
+
+Pure function to get the price in Wei to perform a swap from `inputTokenId` to `outputTokenId` based on [swap price curve](#swap-price-curve).
+
+### `swap`
+
+```solidity
+function swap(uint256 inputTokenId, uint256 outputTokenId) payable 
+```
+
+Swaps the `inputTokenId` from `msg.sender` with the `outputTokenId` from the `swapPool`. Sends the swap fee to the `feeRecipient`.  
+
+Requirements:
+* This contract must be approved to transfer the `nft` with `inputTokenId` on behalf of `msg.sender`.
+* This contract must be approved to transfer the `nft` with `outputTokenId` on behalf of the `swapPool`
+* Swap fee of `swapPrice(inputTokenId, outputTokenId)` Wei must be provided 
```

Comment:
ya nice let's add some `require` statements so we can get better error handling if not approved / under fee. And we could add events as well for each swap to let indexers know a swap occurred.

Reply:
Yep agreed! Didn't go into the actual implementation details here.